src/Metadata/Resource/Factory/ExtractorResourceMetadataFactory.php

```diff
@@ -87,16 +87,23 @@ private function handleNotFound(?ResourceMetadata $parentPropertyMetadata, strin
     }
 
     /**
-     * Creates a new instance of metadata if the property is not already set.
+     * Update resource metadata if new config is given
      */
     private function update(ResourceMetadata $resourceMetadata, array $metadata): ResourceMetadata
     {
         foreach (['shortName', 'description', 'iri', 'itemOperations', 'collectionOperations', 'subresourceOperations', 'graphql', 'attributes'] as $property) {
-            if (null === $metadata[$property] || null !== $resourceMetadata->{'get'.ucfirst($property)}()) {
+            $propertyData = $metadata[$property];
+            $parentPropertyData = $resourceMetadata->{'get'.ucfirst($property)}();
+
+            if (null === $propertyData) {
                 continue;
             }
 
-            $resourceMetadata = $resourceMetadata->{'with'.ucfirst($property)}($metadata[$property]);
+            if (is_array($parentPropertyData) && is_array($propertyData)) {
+                $propertyData = array_merge($parentPropertyData, $propertyData);
```

Comment:
Should we really merge them though? Which is the use case?

Reply:
@alanpoulain Thank you for checking. I am using Sylius, a e-commerce software that utilise ApiPlatform, it [pre-defined  tons of APIs out of the boxes](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/ApiBundle/Resources/config/api_resources). It is common that developer would like to change or extend its behaviour. 

Whether we should merge them, if I am not mistaken it is already the case for `resource.properties`.

https://github.com/api-platform/core/blob/41842c430e1eaf23b18d72a7cbc4744324347be4/src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php#L110-L116

Reply:
Sure we merge the properties, but merging what's inside the properties seems to be dangerous.
Yes I know Sylius allow to extend its resources a lot, but maybe it would be better to redefine the properties when extending instead?

Reply:
@alanpoulain This pull request aim to allow extending resource metadata such as `itemOperations`, `collectionOperations`, `subresourceOperations`, but resource properties mapping is not handled in this class.

Regarding safety, sorry I don't understand why merging config is danger. Symfony bundle configuration support merging too, I naturally assume api resources could do the same. Without proper merging, if one want to add a new operation to an existing resource, I will have to copy and paste all operations from Sylius' definition. 

Anyway, without this pull request even simple overriding is not possible, the new config is discarded due to line 95's null checking.

https://github.com/api-platform/core/blob/41842c430e1eaf23b18d72a7cbc4744324347be4/src/Metadata/Resource/Factory/ExtractorResourceMetadataFactory.php#L94-L100

I am happy to improve the behaviour if you can tell me what is the ideal behaviour. 

Thanks.

Reply:
Merging configuration may be a danger because you also inherit security configuration or serialization context. I think it's better to be explicit in these cases to prevent unwanted configuration.
However I understand your needs, particularly in a Sylius context. I think Sylius needs to provide a decorator for this factory in its bridge with API Platform instead of having it directly here.

Reply:
@alanpoulain Understood. I have two suggestions:

- Throw warning when operation it is already defined, because silently ignoring new config is as danger as silently removing old config.
- Implement deep merge, make it behave like Symfony's `security.yaml` if this makes it any safer.

Kindly let me know your thought. Thanks!